src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -3600,34 +3635,206 @@ int RGWRados::rewrite_obj(RGWBucketInfo& dest_bucket_info, const rgw_obj& obj, c
+    obj.key << " in bucket " << bucket_info.bucket << dendl;
+
+  // we're using a Read::prepare() to perform a synchronous stat
+  // operation of the head object; fill op's param field with pointers
+  // to data we want
+  RGWRados::Object::Read stat_op(&rados_obj);
+  rgw::sal::Attrs attrset;
+  stat_op.params.attrs = &attrset;
+  uint64_t obj_size;
+  stat_op.params.obj_size = &obj_size;
+  ceph::real_time mtime;
+  stat_op.params.lastmod = &mtime;
+#if 0 // currently not needed
+  uint64_t data_pool_epoch;
+  stat_op.params.epoch = &data_pool_epoch;
+#endif
+
+  ret = stat_op.prepare(y, dpp);
+  if (ret == -ENOENT && is_versioned && has_instance) {
+    // head object does not exist if it's a delete marker; handle here
+    // and return
+    ldpp_dout(dpp, 20) << "INFO: " << __func__ << ": indexing " <<
+      obj.key << " as delete marker" << dendl;
+    ret = set_olh(dpp,
+		  obj_ctx,
+		  bucket_info,
+		  obj,
+		  true, // is delete marker
+		  nullptr, // &metadata
+		  0, // uint64_t olh_epoch,
+		  ceph::real_clock::zero(), // unmod_since,
+		  true, // high_precision_time
+		  y, // optional_yield
+		  nullptr, // zones_trace,
+		  false,  // log_data_change
+		  true); // skip the olh object updaste
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+	": unable to set OLH to delete marker for \"" << obj << "\"; returned: " <<
+	cpp_strerror(-ret) << dendl;
+    }
+
+    return ret;
+  } else if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": unable to complete async stat of \"" << obj << "\"; returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  // data we'll pull from head object xattrs
+  bufferlist acl_bl;
+  std::string etag;
+  std::string content_type;
+  std::string storage_class;
+  bool found_compressed;
+  bufferlist compressed_bl;
+  bool found_manifest = false;
+  bufferlist manifest_bl;
+  bool found_olh_ver = false;
+  bufferlist olh_ver_bl;
+  bool appendable = false;
+  bufferlist part_num_bl;
+
+  {
+    auto& attrs = *stat_op.params.attrs;
+    pop_attr(attrs, RGW_ATTR_ACL, acl_bl);
+    read_attr(attrs, RGW_ATTR_ETAG, etag);
+    read_attr(attrs, RGW_ATTR_CONTENT_TYPE, content_type);
+    read_attr(attrs, RGW_ATTR_STORAGE_CLASS, storage_class);
+    pop_attr(attrs, RGW_ATTR_COMPRESSION, compressed_bl, &found_compressed);
+    pop_attr(attrs, RGW_ATTR_OLH_VER, olh_ver_bl, &found_olh_ver);
+    pop_attr(attrs, RGW_ATTR_APPEND_PART_NUM, part_num_bl, &appendable);
+  }
+
+  // initial values of size, which we can override/refine below based
+  // on manifest and compression
+  uint64_t stored_size = *stat_op.params.obj_size;  // size consumed in cluster
+  uint64_t accounted_size = stored_size;            // size if/when decompressed
+
+  if (found_olh_ver) {
+    // this appears to be a pure OLH (i.e., no data); we won't index
+    // as we index actual objects with data and set the OLH then
```

Comment:
i did a quick test with a "null" version (written while versioning was suspended), and a `rados listxattr` of the head object found the following olh attrs there:
```
user.rgw.olh.idtag                                                             
user.rgw.olh.info                                                              
user.rgw.olh.ver
```

Reply:
Thanks!